arcgis-ios-sdk-samples/Maps/Map reference scale/MapReferenceScaleSettingsViewController.swift

```diff
@@ -0,0 +1,226 @@
+    func mapReferenceScaleSettingsViewControllerDidFinish(_ controller: MapReferenceScaleSettingsViewController)
+}
+
+/// A view controller that provides an interface for adjusting the reference
+/// scale and scale of a map.
+class MapReferenceScaleSettingsViewController: UITableViewController {
+    /// The map whose settings should be adjusted.
+    var map: AGSMap!
+    /// The scale of the map. The default it `0`.
+    var mapScale = 0.0
+    /// The delegate of the view controller.
+    weak var delegate: MapReferenceScaleSettingsViewControllerDelegate?
+    
+    @IBOutlet weak var referenceScaleLabel: UILabel!
+    @IBOutlet weak var referenceScalePickerView: UIPickerView!
+    @IBOutlet weak var layersCell: UITableViewCell!
+    
+    @IBOutlet weak var mapScaleLabel: UILabel!
+    @IBOutlet weak var setMapScaleToReferenceScaleButton: UITableViewCell!
+    
+    /// The formatter used to generate strings from scale values.
+    private let scaleFormatter: NumberFormatter = {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        numberFormatter.maximumFractionDigits = 0
+        return numberFormatter
+    }()
+    /// The scale values displayed in the reference scale picker view.
+    private let possibleReferenceScales = [500_000, 250_000, 100_000, 50_000]
+    
+    /// Returns a string containing the formatted value of the provided scale.
+    ///
+    /// - Parameter scale: A scale value.
+    /// - Returns: A string.
+    func string(fromScale scale: Double) -> String {
+        return String(format: "1:%@", scaleFormatter.string(from: scale as NSNumber)!)
+    }
+    
+    /// Called in response to the Done button being tapped.
+    @IBAction func done() {
+        delegate?.mapReferenceScaleSettingsViewControllerDidFinish(self)
+    }
+    
+    /// Indicates whether the reference scale picker is currently hidden.
+    private var referenceScalePickerHidden = true
+    
+    /// Toggles visisbility of the reference scale picker.
+    func toggleReferenceScalePickerVisibility() {
+        tableView.performBatchUpdates({
+        if referenceScalePickerHidden {
+            referenceScaleLabel.textColor = view.tintColor
+            tableView.insertRows(at: [.referenceScalePicker], with: .fade)
+            referenceScalePickerHidden = false
+        } else {
+            referenceScaleLabel.textColor = nil
+            tableView.deleteRows(at: [.referenceScalePicker], with: .fade)
+            referenceScalePickerHidden = true
+        }
+        }, completion: nil)
+    }
+    
+    /// The observer of the reference scale of the map.
+    private var referenceScaleObserver: NSObjectProtocol?
+    /// The observer of the scale of the map.
+    private var scaleObserver: NSObjectProtocol?
+    
+    // MARK: UIViewController
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        precondition(map != nil)
```

Comment:
If the client does not set a map, will this `precondition` crash the app?

Reply:
Correct, just like if they system didn't set one of the IB outlets. Both the data model and the views are needed for this view controller to work properly. I'd love to be able to use an initializer (something like `MapReferenceScaleSettingsViewController.init(map:)`), but that isn't possible with a storyboard. It is possible with a XIB, but XIBs do not support static cells.

Since I made the map an IUO, a crash would happen anyway the first time it was accessed before it was set. I just wanted to make it explicitly clear that the view controller expects the map to be set before being presented. Could I have made the map just an optional? Yes. And there would be a lot more question marks.